storage_controller/src/service.rs

```diff
@@ -2376,61 +2376,75 @@ impl Service {
-                self.config.jwt_token.as_deref(),
-            );
-            // TODO: this, like many other places, requires proper retry handling for 503, timeout: those should not
-            // surface immediately as an error to our caller.
-            let status = client.tenant_delete(tenant_shard_id).await.map_err(|e| {
-                ApiError::InternalServerError(anyhow::anyhow!(
-                    "Error deleting shard {tenant_shard_id} on node {node}: {e}",
-                ))
-            })?;
-            tracing::info!(
-                "Shard {tenant_shard_id} on node {node}, delete returned {}",
-                status
-            );
-            if status == StatusCode::ACCEPTED {
-                any_pending = true;
-            }
+        if let Err(e) = self.await_waiters(detach_waiters, RECONCILE_TIMEOUT).await {
+            // Failing to detach shouldn't hold up deletion, e.g. if a node is offline we should be able
+            // to use some other node to run the remote deletion.
+            tracing::warn!("Failed to detach some locations: {e}");
         }
 
-        if any_pending {
-            // Caller should call us again later.  When we eventually see 404s from
-            // all the shards, we may proceed to delete our records of the tenant.
-            tracing::info!(
-                "Tenant {} has some shards pending deletion, returning 202",
-                tenant_id
-            );
-            return Ok(StatusCode::ACCEPTED);
+        let locations = shard_ids
+            .into_iter()
+            .map(|s| (s, node.clone()))
+            .collect::<Vec<_>>();
+        let results = self.tenant_for_shards_api(
+            locations,
+            |tenant_shard_id, client| async move { client.tenant_delete(tenant_shard_id).await },
+            1,
+            3,
+            RECONCILE_TIMEOUT,
+            &self.cancel,
+        )
+        .await;
+        for result in results {
+            match result {
+                Ok(StatusCode::ACCEPTED) => {
+                    // This could happen if we failed detach above, and hit a pageserver where the tenant
+                    // is still attached: it will accept the deletion in the background
+                    tracing::warn!(
+                        "Unexpectedly still attached on {}, client should retry",
+                        node
+                    );
+                    return Ok(StatusCode::ACCEPTED);
+                }
+                Ok(_) => {}
+                Err(mgmt_api::Error::Cancelled) => {
+                    return Err(ApiError::ShuttingDown);
+                }
+                Err(e) => {
+                    // This is unexpected: remote deletion should be infallible, unless the object store
+                    // at large is unavailable.
+                    tracing::error!("Error deleting via node {}: {e}", node);
+                    return Err(ApiError::InternalServerError(anyhow::anyhow!(e)));
```

Comment:
But it can happen that the "arbitrary" node is unavailable.

What's the rationale behind doing the retries (`tenant_for_shards_api`) above against the same node instead of a different node?

Anyway, if we return the 503 here, what happens?
cplane reports stuck project and it will retry, I suppose? 
That's ok for now, still would be better to try another node each `tenant_for_shards_api`-retry so we handle node failure transparent to cplane.
Can be a follow-up though.

Reply:
>But it can happen that the "arbitrary" node is unavailable.

The reason we call into the scheduler to pick the node is that the scheduler will prefer nodes that are online.

> What's the rationale behind doing the retries (tenant_for_shards_api) above against the same node instead of a different node?

Same as any API call: to cover all the cases where the node is really available but something else weird happened like a network hiccup.  That said, for these cases where we're using "arbitrary available node", it would make sense for the retry loop to try different nodes.

> Anyway, if we return the 503 here, what happens?

If we return an error here, the caller should try again, and we'll eventually pick a node that's available (background heartbeats will figure out if a node is unavailable, and eventually our call into the scheduler will give us an available node).